File: 27_friend_pair_number/27_friendly_pair_number.dart
import 'dart:io';

// Friendly number - two numbers for which  sum of their factor are equal to every other number are called as friendly pai number

List<int> properFactors(int num){
  List<int> factors = [];
  for (int i = 1; i <= (num ~/2); i++){
    if(num % i == 0){
      factors.add(i);
    }
  }
  return factors;
}

int factorSum(List<int> fact){
  int sum = 0;
  for (int i =0; i<fact.length; i++){
    sum+=fact[i];
  }
  return sum;
}

bool isFirendly(num1, num2){
  if (num1 == num2){
    return false;
  }
  int sum1 = factorSum(properFactors(num1));
  int sum2 = factorSum(properFactors(num2));
  return (num1 == sum2 && num2 == sum1);
}

void main(){
  try{
    while(true){
      stdout.write("Enter a positive integer value : ");
      int num1 = int.parse(stdin.readLineSync()!);

      stdout.write("Enter a positive integer value : ");
      int num2 = int.parse(stdin.readLineSync()!);

      int maxLimit = 1000;

      if (num1 > maxLimit || num2 > maxLimit) {
        print("Enter a lesser value than 10000");
      } else if (num1 <= 0 || num2 <= 0) {
        print("Enter a valid integer value : ");
      } else {
        bool result = isFirendly(num1, num2);
        if(result){
          print("$num1 and $num2 are friendly number.");
        }
        else{
          print("$num1 and $num2 are not friendly number.");
        }
      }

    stdout.write("Do you want to continue the program (y/n) : ");
      String choice = stdin.readLineSync()!.toLowerCase();
      if (choice == 'n') {
        print("Program finished!");
        break;
      }
    }
  } on FormatException catch (e) {
    print("Invalid input : $e");
  } on UnsupportedError catch (e) {
    print("Divide by zero exception : $e");
  } on Exception catch (e) {
    print("Exception caught : $e");
  }

}

File: 27_friend_pair_number/27_friendly_pair_number.py
# Python program to check whether the pair of given number is a friendly pair or not
# Friendly pair - two two number should be a perfect number then it si friendly pair

# method 1 - using iteration
# 1st - range(1,n+1)
# 2nd - range(1,n//2+1)  --- more fesible (complexity reduces by half times)
# 3rd - range(1,sqrt(n))  --- most feasible (complexity reduces by square root times)

from math import sqrt

n,m=map(int,input("Enter the space separated number :").split(" "))
result1 = result2 = False
for i in n,m:
    sum=0
    for j in range(1,i//2+1):
        if i%j==0:
            sum=sum+j
    if i==sum:
        if i==n:
            result1=True
        elif i==m:
            result2=True
if result1 and result2:
    print("It is a Friendly pair")
else:
    print("It is not a friendly pair")

# method 2 - using functions and recurssion
a,b=map(int,input().split())
def fun(m):
    global sum
    sum=0
    def rec(n):
        if n==m//2+1:
            return sum
        if m%n==0:
            sum=sum+n
        rec(n+1)
    rec(1)
if fun(a)==a and fun(b)==b:
    print("It is a Friendly Pair")
else:
    print("It is not a Friendly Pair")

# method. 3 - using module based approach

def proper_factor(num):
    factor_list = [ i for i in range(1, num//2+1) if num%i==0]
    return factor_list

def is_friendly(num1, num2):
    if num1 == num2:
        return False
    sum1 = sum(proper_factor(num1))
    sum2 = sum(proper_factor(num2))
    return (num1==sum2 and num2==sum1)

try:
    while(True):
        num1 = int(input("Enter positive integer value : "))
        num2 = int(input("Enter positive integer value : "))
        limit  = 1000
        if (num1 <= 0 or num2 <= 0):
            print("Enter positiveinteger value")
        elif (num1 > limit or num2 > limit):
            print("Input too large")
        else:
            result = is_friendly(num1, num2)
            if(result):
                print(F"{num1}, {num2} are Friendly Pair")
            else:
                print(F"{num1}, {num2} are not Friendly Pair")
        choice = input("Do you want to continue the program (y/n) : ")
        if(choice == "n"):
            print("Program finished!")
            break

except ValueError as e:
    print("Invalid Input : {e}")

except ZeroDivisionError as e:
    print(f"Zero division exception : {e}")

except Exception as e:
    print(f"Exception caught : {e}")


